[FX][1/2] Separate inputs from the Node list

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45227 [FX][WIP] Mutable Graph APIs
* #45545 [FX][2/2] Make output a node
* **#45245 [FX][1/2] Separate inputs from the Node list**

First out of two PRs to make the handling of inputs/outputs in FX match more closely to the TorchScript IR, and to facilitate better APIs for mutable conversion.

This PR separates out placeholder nodes from the rest of the node list, storing them in a separate `_inputs` list. These nodes are still available via the `.inputs` property.

Differential Revision: [D23890345](https://our.internmc.facebook.com/intern/diff/D23890345)

Commit:
[FX][WIP] Make return a node

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Make return a node"


Differential Revision: [D23890345](https://our.internmc.facebook.com/intern/diff/D23890345)

[ghstack-poisoned]

zdevito commented on torch/fx/graph.py:
```diff
@@ -67,7 +67,7 @@ def map_arg(a: Argument, fn: Callable[[Node], Argument]) -> Argument:
 class Graph:
     def __init__(self):
-        self._nodes : List[Node] = []
+        self._nodes : List[Node] = [Node(self, 'return', 'return', 'return', (), {})]
```
This is inconsistent with how placeholders are handled. A lot of the later complexity is related to the fact that the return node is inserted into the graph first. Why not insert it when `graph.output` is called (i.e. treat output like placeholder in that it inserts the node last).

zdevito commented on torch/fx/graph.py:
```diff
@@ -148,8 +152,14 @@ def node_copy(self, node: Node, arg_transform: Callable[[Node], Argument] = lamb
+        assert self._nodes[-1].op == 'return'
+        self._nodes[-1].args = result
+
+    @property
```
`result` is redundant now with the presence of a return node. The code in the if statement handling nodes should be the thing handling the result otherwise you end up having two places where everything handles stuff.

zdevito commented on torch/fx/node.py:
```diff
@@ -25,8 +25,8 @@ def __init__(self, graph: 'Graph', name: str, op: str, target: Target,
         self.name = name  # unique name of value being created
-        assert op in ['placeholder', 'call_method', 'call_module', 'call_function', 'get_attr']
-        self.op = op  # the kind of operation = placeholder|call_method|call_module|call_function|get_attr
+        assert op in ['placeholder', 'call_method', 'call_module', 'call_function', 'get_attr', 'return']
```
the node that creates the "return" node is called ".output"  that seems inconsistent. Rename return to output?

Commit:
Update on "[FX][WIP] Make return a node"


This makes things like iterating over all uses more consistent and makes returns look like function arguments (placeholders)

Things I don't like:
- Needing to explicitly skip 'return' nodes in a few places
- People can't do nodes[-1] to get the last node anymore

Differential Revision: [D23890345](https://our.internmc.facebook.com/intern/diff/D23890345)

[ghstack-poisoned]

zdevito reviewed (commented):
I am generally for either (1) both placeholders/output are in the node list and considered operators, or (2) neither of them are. This patch seems to put things into a middle state, where placeholders are added via append, but output is added at the beginning. Placeholders have no special accessor, but the output node is forced to be singluar and does have a single accessor. So the change doesn't seem consistent.

Commit:
Update on "[FX][WIP] Make return a node"


This makes things like iterating over all uses more consistent and makes returns look like function arguments (placeholders)

Things I don't like:
- Needing to explicitly skip 'return' nodes in a few places
- People can't do nodes[-1] to get the last node anymore

Differential Revision: [D23890345](https://our.internmc.facebook.com/intern/diff/D23890345)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Make return a node"


This makes things like iterating over all uses more consistent and makes returns look like function arguments (placeholders)

Things I don't like:
- Needing to explicitly skip 'return' nodes in a few places
- People can't do nodes[-1] to get the last node anymore

Differential Revision: [D23890345](https://our.internmc.facebook.com/intern/diff/D23890345)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Make return a node"


This makes things like iterating over all uses more consistent and makes returns look like function arguments (placeholders)

Things I don't like:
- Needing to explicitly skip 'return' nodes in a few places
- People can't do nodes[-1] to get the last node anymore

Differential Revision: [D23890345](https://our.internmc.facebook.com/intern/diff/D23890345)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Make return a node"


This makes things like iterating over all uses more consistent and makes returns look like function arguments (placeholders)

Things I don't like:
- Needing to explicitly skip 'return' nodes in a few places
- People can't do nodes[-1] to get the last node anymore

Differential Revision: [D23890345](https://our.internmc.facebook.com/intern/diff/D23890345)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Make return a node"


This makes things like iterating over all uses more consistent and makes returns look like function arguments (placeholders)

Things I don't like:
- Needing to explicitly skip 'return' nodes in a few places
- People can't do nodes[-1] to get the last node anymore

Differential Revision: [D23890345](https://our.internmc.facebook.com/intern/diff/D23890345)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Make return a node"


This makes things like iterating over all uses more consistent and makes returns look like function arguments (placeholders)

Things I don't like:
- Needing to explicitly skip 'return' nodes in a few places
- People can't do nodes[-1] to get the last node anymore

Differential Revision: [D23890345](https://our.internmc.facebook.com/intern/diff/D23890345)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Make return a node"


This makes things like iterating over all uses more consistent and makes returns look like function arguments (placeholders)

Things I don't like:
- Needing to explicitly skip 'return' nodes in a few places

Differential Revision: [D23890345](https://our.internmc.facebook.com/intern/diff/D23890345)

[ghstack-poisoned]

Commit:
Update on "[FX][WIP] Make return a node"


This makes things like iterating over all uses more consistent and makes returns look like function arguments (placeholders)

Things I don't like:
- Needing to explicitly skip 'return' nodes in a few places

Differential Revision: [D23890345](https://our.internmc.facebook.com/intern/diff/D23890345)

[ghstack-poisoned]

Commit:
Update on "[FX][1/2] Separate inputs from the Node list"


First out of two PRs to make the handling of inputs/outputs in FX match more closely to the TorchScript IR, and to facilitate better APIs for mutable conversion.

This PR separates out placeholder nodes from the rest of the node list, storing them in a separate `_inputs` list. These nodes are still available via the `.inputs` property.

Differential Revision: [D23890345](https://our.internmc.facebook.com/intern/diff/D23890345)

[ghstack-poisoned]

Commit:
Update on "[FX][1/2] Separate inputs from the Node list"


First out of two PRs to make the handling of inputs/outputs in FX match more closely to the TorchScript IR, and to facilitate better APIs for mutable conversion.

This PR separates out placeholder nodes from the rest of the node list, storing them in a separate `_inputs` list. These nodes are still available via the `.inputs` property.

Differential Revision: [D23890345](https://our.internmc.facebook.com/intern/diff/D23890345)

[ghstack-poisoned]

Commit:
Update on "[FX][1/2] Separate inputs from the Node list"


First out of two PRs to make the handling of inputs/outputs in FX match more closely to the TorchScript IR, and to facilitate better APIs for mutable conversion.

This PR separates out placeholder nodes from the rest of the node list, storing them in a separate `_inputs` list. These nodes are still available via the `.inputs` property.

Differential Revision: [D23890345](https://our.internmc.facebook.com/intern/diff/D23890345)

[ghstack-poisoned]

Commit:
Update on "[FX][1/2] Separate inputs from the Node list"


First out of two PRs to make the handling of inputs/outputs in FX match more closely to the TorchScript IR, and to facilitate better APIs for mutable conversion.

This PR separates out placeholder nodes from the rest of the node list, storing them in a separate `_inputs` list. These nodes are still available via the `.inputs` property.

Differential Revision: [D23890345](https://our.internmc.facebook.com/intern/diff/D23890345)

[ghstack-poisoned]

Commit:
Update on "[FX][1/2] Separate inputs from the Node list"


First out of two PRs to make the handling of inputs/outputs in FX match more closely to the TorchScript IR, and to facilitate better APIs for mutable conversion.

This PR separates out placeholder nodes from the rest of the node list, storing them in a separate `_inputs` list. These nodes are still available via the `.inputs` property.

Differential Revision: [D23890345](https://our.internmc.facebook.com/intern/diff/D23890345)

[ghstack-poisoned]
